### ISSUE TITLE
Fusion with ipyopt?

### ISSUE THREAD

**Issue by g-braeunlich** [mechmotum/cyipopt#95]
Dear cyipopt devs
As the main (and currently only) developer of [ipyopt](https://gitlab.com/g-braeunlich/ipyopt/), I just have been contacted by a dev of [zlib](https://github.com/zfit/zfit), see [issue](https://gitlab.com/g-braeunlich/ipyopt/-/issues/1#note_513738029).
He is asking whether ipyopt is coordinated in any way with cyipopt and was suggesting to join the projects.

Of course, I would support collaborating in any way - given that you would agree.

However from what I can see, our projects are very different in their approaches:
cyipopt: cython binding to the C API
ipyopt: almost ported to a pybind11 binding to the C++ API (see the [cpp](https://gitlab.com/g-braeunlich/ipyopt/-/tree/cpp) and [pybind11](https://gitlab.com/g-braeunlich/ipyopt/-/tree/pybind11) branches of ipyopt).

So question:
What do you think? Would a collaboration / merge would make sense?
From my side: I would be open to almost everything (license - as far as I am legally free here, code repo hoster, project owner).

**Comment by moorepants** [comment]
We conversed on email a bit so I'll post my response here:

Gerhard,

Thanks for reaching out. It would certainly be nice if there were a single Python binding to IPOPT so that effort is not duplicated and end users have the best end product. Around 7 years ago I was looking for a way to use IPOPT in Python and I think I found https://github.com/alanfalloon/pyipopt and https://bitbucket.org/amitibo/cyipopt. I was able to get the later working and usable more quickly for whatever reason. A handful of people have kept cyipopt alive over the years. It is a Cython binding to the C API of IPOPT that attempts to expose a Pythonic API and includes a small number of bells and whistles. I'm personally not too familiar with C++, so Cython + C has been more approachable.

I'm open to discussion about merging in some way. One our end some of the important things for us have been:

- Cython based so we can fully control the Python API
- Keeping it simple with the C API
- Easy installations on Linux, Mac, and Windows via conda and conda forge
- Including a scipy-like interface
- Backwards compatibility and keeping things stable
- Not doing much more than providing a way to define an NLP problem

Happy to hear any proposals.

**Comment by g-braeunlich** [comment]
Thanks for your feedback!
The last 4 points in your list are basically also my goals (I did not think about the option of having a scipy-like interface, but I really like the idea!).
My considerations to the C interface:
As you can see here: https://github.com/coin-or/Ipopt/blob/master/src/Interfaces/IpStdCInterface.h
IPOpt's C interface is an additional layer on top of their C++ core, which adds some overhead (they are using IpStdInterfaceTNLP which stores pointers to the C callbacks and calls them in its own callbacks (e.g. eval_f)).
Moreover it seems to me that the C interface is not as generic as the plain C++ version. For example in the C interface, you can only define some constant scaling parameters:
https://github.com/coin-or/Ipopt/blob/master/src/Interfaces/IpStdCInterface.h#L280

Where as in the C++ version, you have a callback in which you dynamically can change scaling:
https://github.com/coin-or/Ipopt/blob/master/src/Interfaces/IpTNLP.hpp#L211

Thus: I really would like to use the C++ interface directly (AFAIK cython also can wrap C++).

If you would be open to that, here is my proposal:
* I already have the C++ interfacing part - we could use it as a common base.
* We would apply your cython stack, including the scipy interface on top of that
* We would hopefully be able to use your existing packaging mechanism to keep support for condaforge / portability
* In addition to that, I also would bundle my existing pure C++ extension which I could upoload as a wheel to pypi (as I am doing now).
* I also would be interested to port the scipy interface to plain C++. Reason: By using cython, you are basically using a domain specific language which binds your application to the python ecosystem - you cannot (at least I dont know that this would be possible) use this part natively for e.g. C++.

What do you think?

**Comment by stonebig** [comment]
Qs a pip user only, I would dream to have windows binary wheels and recipe from at least one of the two projects

**Comment by moorepants** [comment]
> As a pip user only, I would dream to have windows binary wheels and recipe from at least one of the two projects.

These things are still a bit tough with PIP unfortunately, especially cross platform. Getting the windows builds using the conda stack for windows took us years! If there was a windows wheel, I think there needs to be an ipopt windows wheel first.

**Comment by mayou36** [comment]
That sounds very good to me, the proposal of @g-braeunlich . What I think is _maybe_ a point is the simple interface:
- scipy interface is for free, sure!
- full control sounds quite tempting, indeed. Been playing with the knobs a lot
- however, the interface seems not always optimal with some namings (at least to me) and/or with what is possible to give in. But still, full control sounds good.

(I am the one from [zfit](https://github.com/zfit/zfit), where we're wrapping ipopt into the likelihood minimization library)

**Comment by moorepants** [comment]
@g-braeunlich, I think to have any movement on this someone needs to start drafting a detailed merging plan that would show both how the merged codebases would look and how we support backwards compatibility with both libraries. Also having a master list of our combined needs that give some indication of how those are going to be met with the new codebase, infrastructure, etc.

Our priority here is to do a 1.0 release in the coming months and that will probably take most of our developer volunteer time. If you feel motivated to push this, I think you'll need to provide the merging plan, or at least a draft.

**Comment by g-braeunlich** [comment]
@moorepants OK. Then first lets discuss the following points:
* Where should the merged project live? My preference would be gitlab (just because of ideological reasons - gitlab CE is open source). I also could live with github. But then I would strongly suggest to use gitlab actions instead of travis CI (no 3rd party CI system, no additional accounts needed).
* What should be the name of the merged project? Here I would offer "ipyopt" (we are mainly targeting python, the prefix "cy" might be confusing to some users. This is also just my preference. Open to discussions)
* License? I had to go with BSD 3 because I took over maintainership. Since the whole codebase has now be ported to cpp, I think there should not be much left of the original code, so I would be basically free to switch to whatever license

**Comment by mayou36** [comment]
1. no reason for one or the other from my side. Travis is now paid only, factually, but Github has its own CI now: github actions.
2. I would, from a user side, prefer ipyopt over cyipopt, the former goes more natural
3. I tend towards well-known licenses, so BSD 3 (seems very similar to the Eclipse license) I would go for

Many thanks for the effort!

**Comment by g-braeunlich** [comment]
Sorry, meant to write "github actions" and not "gitlab actions". I corrected it in my above comment

**Comment by moorepants** [comment]
> Where should the merged project live? My preference would be gitlab (just because of ideological reasons - gitlab CE is open source). I also could live with github. But then I would strongly suggest to use github actions instead of travis CI (no 3rd party CI system, no additional accounts needed).

Gitlab is fine with me.

> What should be the name of the merged project? Here I would offer "ipyopt" (we are mainly targeting python, the prefix "cy" might be confusing to some users. This is also just my preference. Open to discussions)

I have not strong preferences for a name. ipyopt is fine.

> License? I had to go with BSD 3 because I took over maintainership. Since the whole codebase has now be ported to cpp, I think there should not be much left of the original code, so I would be basically free to switch to whatever license

I prefer MIT or BSD when I have a choice, but if code is actually merged we need to read the EPL and understand any implications and abide by them.

**Comment by moorepants** [comment]
@brocksam, @bocklund  Please chime in on this discussion if you have opinions.

**Comment by richardotis** [comment]
Gitlab is a critical part of the FOSS ecosystem run by talented people, but most of the scientific Python ecosystem, including CF packaging, lives on GitHub. You will get less integrated crosslinking of relevant GH issues in other projects as they come up, and my experience is that Gitlab projects have worse SEO; for whatever reason, it's a hassle using Google to find that a Gitlab project exists. I think the latter discoverability concern is a huge issue for a project that has only a few (but dedicated! your downstream thanks you!) volunteers and has historically been a big maintenance/packaging burden, especially for Windows.

Regarding the `pip` issue, especially for Windows, I'd suggest looking if `conda press` can meet your needs: https://github.com/conda-incubator/conda-press

Without doing any analysis, I expect that the pyipopt approach of directly using the Python/Numpy API to shuffle data to and from IPOPT is likely to be more performant than the semi-typed Cython approach of cyipopt. My experience using cyipopt is that the overhead of creating the Problem objects and calling `solve` in tight loops is significant, though whether that's Cython/C-API overhead or just IPOPT overhead, I do not know. It'd be great to take the best bits of both codes and bring them together, though I think it's likely to be a more-than-expected amount of work (especially to maintain cross-platform).

Regarding scipy interfaces, you may find this example using `LowLevelCallable` interesting: https://docs.scipy.org/doc/scipy/reference/tutorial/integrate.html#quad-callbacks

**Comment by brocksam** [comment]
> @brocksam, @bocklund Please chime in on this discussion if you have opinions.

1. I agree with @richardotis on GitHub regarding search accessibility and aligning with the scientific Python ecosystem. GitHub is preferred from my perspective as the conda-feedstock would need to live on GitHub and splitting the project across hosting platforms would be clumsy. I'd fully support transferring CI over to GitHub actions (even for cyipopt).
2. I'd be fine with ipyopt as the name, I agree with @g-braeunlich that it's less confusing the cyipopt. But it does raise the question about how backwards compatibility is maintained? Does ipyopt essentially become a completely new project on GitHub and take over the ipyopt on PyPI and conda-forge? If the merged ipyopt gets up and running, do we release a new cyipopt which issues deprecation warnings pointing the user to install and use ipyopt instead?
3. I prefer the MIT license, but anything equally permissible is fine with me.
4. @moorepants makes a very good point that cyipopt's priority right now is the 1.0 release. The proposal for a single unified Python-Ipopt project is nice, but where is the development time and effort coming from? It's going to be a lot of work, with not much immediate benefit (from my personal perspective), so I'm uncertain how much of my time I'd be able to volunteer.
5. I'm on board with @g-braeunlich's proposal of using the C++ API as it exposes all the bells and whistles of Ipopt (which the C API doesn't), plus any performance improvement would be a nice bonus.

**Comment by g-braeunlich** [comment]
1. OK. I take the concerns of @brocksam and @richardotis serious. Therefore Id say github wins over gitlab.
2. 
>     Does ipyopt essentially become a completely new project on GitHub and take over the ipyopt on PyPI and conda-forge? If the merged ipyopt gets up and running, do we release a new cyipopt which issues deprecation warnings pointing the user to install and use ipyopt instead?
I'd say: yes!

3. License: How can we decide this?
4. 
>     4. The proposal for a single unified Python-Ipopt project is nice, but where is the development time and effort coming from? It's going to be a lot of work, with not much immediate benefit (from my personal perspective), so I'm uncertain how much of my time I'd be able to volunteer.
If you want me (and probably others) to join the project, then thats the price! If you dont want to make that investment into the future, well then we can continue working on each fork on our own.
The dev time would come from my side partially. I would offer to (any help welcome)
* Create the new project space
* Put my C++ code into the new repo
* Port my CI to gitlab actions
* Try to get the cython stack running on top of the C++ code

5. Nothing to discuss here

**Comment by mayou36** [comment]
Hey, so this sounds quite good to me.

To 4, I agree: there is initial work (actually of porting and merging code, not rewriting from scratch), but less on the longterm as there will be less overall duplicated code, we won't have a split userbase etc.

I would be quite in favor of this as - honestly - there are things I like from both packages and having them merged seems like a good approach.

The next step would maybe be to create an org and try to get things up there.
Or what do you think?

**Comment by g-braeunlich** [comment]
Hey @mayou36 
From my side, I am currently working on the suggestion of @richardotis (thanks for bringing that to my attention)
i.e. support for `LowLevelCallable`. I almost finished support for `PyCapsules`. This should significantly improve performance.
Once that is done, my next step would be a scipy like interface for ipyopt (goal: API compatibility with cyipopt).

Of course, I am happy to create a new org at any time, we agree on a license (from my side, I dont have a strong preference).

**Comment by g-braeunlich** [comment]
As It turned out while developing my own method (see https://ipyopt.readthedocs.io/en/latest/scipy.html), it is not trivial to arrange ipyopt's goals (providing maximal performance) with a scipy.optimize.mimimize method which is as close to scipy's native methods as cyipopt's interface.
Therefore, I came to the conclusion, that our 2 projects actually have different goals. 
ipyopt will continue its path and now supports PyCapsules (which btw. also could be very interesting for you - PyCapsules can be generated easily with cython: https://ipyopt.readthedocs.io/en/latest/capsules.html).
As a consequence, I decided to close this issue.

I added a link to cyipopt in the README and in the first link above.
Also, in the mean time, @mayou36 founded a new organization https://github.com/py-nonlinopt/ which I joined.
Feel free to join too!
